File: src/circulardoublylinkedlist.cpp
#include <iostream>
#include <cassert>
#include <climits>

#include "../include/circulardoublylinkedlist.h"

// Default constructor
CircularDoublyLinkedList::CircularDoublyLinkedList()
    : head {NULL}, tail {NULL}, count {0} {
}

// Evaluate if the singly link list has any element
bool CircularDoublyLinkedList::isEmpty() const {
    return head == NULL;
}

// Desctructor
CircularDoublyLinkedList::~CircularDoublyLinkedList() {
    clear();
}

// Remove all nodes in the singly link list
void CircularDoublyLinkedList::clear() {

    tail->next = NULL;

    for(TwoWayNode *ptr; !isEmpty();) {
        ptr = head->next;
        delete head;
        head = ptr;
    }
    count = 0;
}

// Get the number of list elements
int CircularDoublyLinkedList::getCount() {
    return count;
}

// Get the value of the head element
int CircularDoublyLinkedList::getHeadData() {
    return head->value;
}

// Get the value of the tail element
int CircularDoublyLinkedList::getTailData() {
    return tail->value;
}

// Get head address
TwoWayNode *CircularDoublyLinkedList::getPtrHead() {
    return head;
}

// Get tail address
TwoWayNode *CircularDoublyLinkedList::getPtrTail() {
    return tail;
}

// Verify if the singly link list has elements
void CircularDoublyLinkedList::emptyness() {
    assert(!isEmpty());
}

// Print all elements the sibgly link list
void CircularDoublyLinkedList::display() {

    using std::cout;

    // If there are not elements
    if (isEmpty()) {
        cout << "The circular doubly linked list is empty.\n";
        return;
    }

    TwoWayNode *ptr {head};
    cout << "Elements in the circular linked list: \n";

    for (int i {0}; i < count; ++i) {
        cout << ptr->value << " ";
        ptr = ptr->next;
    }

    cout << "\n";
}

void CircularDoublyLinkedList::verifyIndex(int index) {
    assert(index >= 0 && index < count);
}

// * Adds new elements

// Adds a new element at beginning of singly link list
void CircularDoublyLinkedList::push(int value) {

    TwoWayNode *newNode = new TwoWayNode(value);

    // When is empty the tail is the new element
    if (head == NULL) {
        head = tail = newNode;
        head->prev = tail;
        head->next = head;
    } else {
        head->prev = newNode;
        newNode->next = head;
        newNode->prev = tail;
        head = newNode;
    }

    ++count;
}

// Adds a new element at end of singly link list
void CircularDoublyLinkedList::append(int value) {

    TwoWayNode *newNode = new TwoWayNode(value);

    // If list is not empty
    if (head == NULL) {
        head = newNode;
        head->prev = tail;
        head->next = head;
    } else {
        tail->next = newNode;
        newNode->prev = head;
        newNode->next = head;
        tail = newNode;
    }

    ++count;
}

// Adds a new element at a given index in the singly link list
void CircularDoublyLinkedList::insert(int index, int value) {

    // Verify if the singly link list has elements
    emptyness();

    // Verification for no more elements in the singly link list
    verifyIndex(index);

    // Add new element
    if (index == 0)
        push(value);
    else if (index == count-1)
        append(value);
    else {

        TwoWayNode *ptr {head};

        for (int i {0}; i < index-1; ++i)
            ptr = ptr->next;

        ptr->next = new TwoWayNode(value, ptr, ptr->next);
        ptr->next->next->prev = ptr->next;
        ++count;
    }
}

// Adds a new element in a sorted singly link list
void CircularDoublyLinkedList::insertSorted(int value) {

    // Existance of elements in the singly link list
    emptyness();

    // Check if the new value is less than head
    if (head->value >= value)
        push(value);
    else if (tail->value <= value)
        append(value);
    else {

        TwoWayNode *ptr {head->next};

        while (ptr != tail && ptr->value <= value && !(ptr->value > value))
            ptr = ptr->next;

        ptr->prev->next = new TwoWayNode(value, ptr->prev, ptr);
        ptr->prev = ptr->prev->next;
        ++count;
    }
}

// Adds a new element before the first apperance of the first ocurrance of a given element
void CircularDoublyLinkedList::beforeValue(int previousValue, int value) {

    emptyness();

    // Check the first element
    if (head->value == previousValue)
        push(value);
    else {

        TwoWayNode *ptr {head};

        while (ptr != tail && ptr->value != previousValue)
            ptr = ptr->next;

        if (ptr->value == previousValue) {
            ptr->prev->next = new TwoWayNode(value, ptr->prev, ptr);
            ptr->prev = ptr->prev->next;
            ++count;
        }
    }
}

// Add a new element after the first apperance of a given value
void CircularDoublyLinkedList::afterValue(int afterValue, int value) {

    emptyness();

    // There is only one element
    if (count == 1 && head->value == afterValue)
        append(value);
    else {

        TwoWayNode *ptr {head};

        while (ptr != tail && ptr->value != afterValue)
            ptr = ptr->next;

        if (ptr == tail && ptr->value == afterValue)
            append(value);
        else if (ptr->value == afterValue) {
            ptr->next = new TwoWayNode(value, ptr, ptr->next);
            ptr->next->next->prev = ptr->next;
            ++count;
        }
    }
}

// * Search methods

// Apply the linear search to the singly link list
TwoWayNode *CircularDoublyLinkedList::search(int value) {

    emptyness();

    // Check the extremes
    if (head->value == value)
        return head;
    else if (tail->value == value)
        return tail;

    TwoWayNode *ptr {head}, *temp {nullptr};

    for (int i {0}; i < count; ++i) {

        if (ptr->value == value) {

            temp = ptr;
            ptr->prev->next = ptr->next;
            ptr->next->prev = ptr->prev;

            ptr->next = head;
            ptr->prev = tail;
            head = ptr;

            return temp;
        }

        ptr = ptr->next;
    }

    // In case its not found
    return NULL;
}

// * Methods of deletion

int CircularDoublyLinkedList::shift() {

    // Check if the singly link list has elements
    emptyness();

    TwoWayNode *ptr {head};
    int x {ptr->value};

    if (count == 1) {
        clear();
        return x;
    }

    head = head->next;
    head->prev = tail;
    tail->next = head;
    delete ptr;
    --count;

    return x;
}

// Removes the last element in the list
int CircularDoublyLinkedList::pop() {

    emptyness();

    TwoWayNode *ptr {tail};
    int x {ptr->value};

    // If there is just one element
    if (count == 1) {
        clear();
        return x;
    }

    tail = tail->prev;
    tail->next = head;
    delete ptr;
    --count;

    return x;
}

// Delete an element by a given index
int CircularDoublyLinkedList::Delete(int index) {

    emptyness();
    verifyIndex(index);

    if (count == 1) {
        int x {head->value};
        clear();
        return x;
    }

    // Verify the extremes
    if (index == 0)
        return shift();
    else if (index == count-1)
        return pop();
    else {

        TwoWayNode *ptr {head};

        for (int i {0}; i < index; ++i)
            ptr = ptr->next;

        int x {ptr->value};
        ptr->prev->next = ptr->next;
        ptr->next->prev = ptr->prev;
        delete ptr;
        --count;

        return x;
    }
}

// Remove duplicates
void CircularDoublyLinkedList::removeDuplicatesSorted() {

    emptyness();

    TwoWayNode *ptr {head->next}, *temp {nullptr};

    while (ptr !=tail) {

        if (ptr->prev->value != ptr->value)
            ptr = ptr->next;
        else {
            temp = ptr->next;
            ptr->prev->next = temp;
            temp->prev = ptr->prev;
            delete temp;
            --count;
            ptr = temp;
        }
    }

    if (ptr == tail && ptr->prev->value == ptr->value)
        pop();
}

// Delete the first apperance of the given value
int CircularDoublyLinkedList::remove(int value) {

    emptyness();

    // In case its the first element
    if (head->value == value)
        return shift();

    TwoWayNode *ptr {head};
    int x {INT_MIN};

    while (ptr != tail && ptr->value != value)
        ptr = ptr->next;

    if (ptr == tail && ptr->value == value)
        return pop();
    else if (ptr != tail && ptr->value == value) {
        x = ptr->value;
        ptr->prev->next = ptr->next;
        ptr->next->prev = ptr->prev;
        delete ptr;
        --count;

        return x;
    }

    return x;
}

// * Methods of sorting

// Check if the singly link list is sorted
bool CircularDoublyLinkedList::checkSort() {

    emptyness();

    TwoWayNode *ptr {head};
    int x {INT_MIN};

    for (int i {0}; i < count; ++i) {
        if (ptr->value < x)
            return false;

        x = ptr->value;
        ptr = ptr->next;
    }

    return true;
}

void CircularDoublyLinkedList::reverse() {

    emptyness();

    TwoWayNode *ptr {head}, *prevNode {nullptr}, *nextNode {nullptr};
    tail = ptr;

    for (int i {0}; i < count; ++i) {
        nextNode = ptr->next;
        ptr->next = prevNode;
        ptr->prev = nextNode;
        prevNode = ptr;
        ptr = nextNode;
    }

    head = prevNode;
    head->prev = tail;
    tail->next = head;
}

// Check if the singly link list is looped otherwise is linear
bool CircularDoublyLinkedList::isLoop() {

    emptyness();

    TwoWayNode *ptr {head}, *temp {head};

    do {
        ptr = ptr->next;
        temp = temp->next;
        temp = (temp) ? temp->next : NULL;
    } while (ptr && temp && ptr != temp);

    return (ptr == temp) ? true : false;
}

// * Methods of update

// Replace all values or the first apperance of a given value with a new one
void CircularDoublyLinkedList::replace(int oldValue, int value, bool entire) {

    emptyness();

    TwoWayNode *ptr {head};

    for (int i {0}; i < count; ++i) {
        // Replace the value
        if (ptr->value == oldValue) {
            ptr->value = value;

            // Stop in case the entire value is false
            if (!entire)
                break;
        }

        ptr = ptr->next;
    }
}

// Replace an value with a new value by a given index
void CircularDoublyLinkedList::update(int index, int value) {

    // Existance of the elements in the list
    emptyness();
    // Check the range of the index
    verifyIndex(index);

    TwoWayNode *ptr {head};

    for (int i {0}; i < index; ++i)
        ptr = ptr->next;

    ptr->value = value;
}

File: python/circulardoublylinkedlist.py
from sys import maxsize
from dataclasses import dataclass
from node import TwoWayNode
from typing import Generator


@dataclass
class CircularDoublyLinkedList:
    pass


    def __post_init__(self):
        self.__head: TwoWayNode = None
        self.__tail: TwoWayNode = None
        self.__count: int = 0


    @property
    def head(self) -> TwoWayNode:
        """Gets the property of the head of the circular doubly linked list."""
        return self.__head


    @property
    def tail(self) -> TwoWayNode:
        """Gets the property of the tail of the circular doubly linked list."""
        return self.__tail


    @property
    def count(self) -> int:
        """Returns the number of items in the circular doubly linked list."""
        return self.__count


    def __len__(self) -> int:
        """Returns the count of nodes in the circular doubly linked list."""
        return self.__count


    def __str__(self) -> str:
        """Returns the string representation of the circular doubly linked list."""
        return f'CircularDoublyLinkedList(head={self.head}, tail={self.tail}, count={self.__count})'


    def __existance_of_elements(self):
        """Asserts if there is any node in the list."""
        ## Evaluate the existence of the single linked list
        assert self.head, 'There are no elements in the doubly linked list. Use the add or append method.'


    def __iter__(self, reverse: bool) -> Generator:
        """Iterates through the all nodes of the circular doubly linked list."""
        ## If there is not head
        self.__existance_of_elements()

        ## Iterate until the last element of the list
        probe: TwoWayNode = self.head if not reverse else self.tail

        for _ in range(self.count):
            yield probe
            probe = probe.next if not reverse else probe.prev


    def display(self, reverse: bool = False):
        """Display all nodes values of the circular doubly linked list."""
        self.__existance_of_elements()

        for node in self.__iter__(reverse):
            print(node.value)


    def __verify_index(self, index: int):
        """Verifies the index of the input."""
        if not(-1 <= index < self.count):
            raise IndexError(f'The index value: {index} is out of range in the circular doubly linked list.')


    def push(self, value: int):
        """Adds a node to the begging of the circular doubly linked list."""
        
        new_node: TwoWayNode = TwoWayNode(value)
        
        if not self.head:
            self.__head = self.__tail = new_node
            self.__head.prev = self.tail
            self.__head.next = self.head
        else:
            self.__head.prev = new_node
            new_node.next = self.head
            new_node.prev = self.tail
            self.__head = new_node

        self.__count += 1
        

    def append(self, value: int):
        """Appends node to the end of the circular doubly linked list."""

        new_node: TwoWayNode = TwoWayNode(value)

        if not self.head:
            self.__head = self.__tail = new_node
            self.__head.prev = self.tail
            self.__head.next = self.head
        else:
            self.__tail.next = new_node
            new_node.prev = self.tail
            new_node.next = self.head
            self.__tail = new_node

        self.__count += 1


    def insert(self, index: int, value: int):
        """Inserts a new node in function of the index."""
        self.__existance_of_elements()
        ## Verify if list is empty
        self.__verify_index(index)

        ## Inserts on the index
        if index == 0:
            ## Add the beginning
            self.push(value)
        elif index in [-1, self.count-1]:
            ## Add the end of the single linked list
            self.append(value)
        else:
            probe: TwoWayNode = self.head

            for _ in range(index-1):
                probe = probe.next

            probe.next = TwoWayNode(value, probe.next, probe)
            probe.next.next.prev = probe.next
            self.__count += 1


    def insert_sorted(self, value: int):
        """Insert a new node into the sorted list."""
        self.__existance_of_elements()

        if self.head.value >= value:
            self.push(value)
        elif self.tail.value <= value:
            self.append(value)
        else:
            probe: TwoWayNode = self.head.next

            while probe != self.tail and probe.value <= value and not(probe.value > value):
                probe = probe.next

            probe.prev.next = TwoWayNode(value, probe, probe.prev)
            probe.prev = probe.prev.next
            self.__count += 1


    def before_value(self, existing_value: int, value: int):
        """Inserts a new node before the first appearance of a given value."""
        ## Evaluate if the single linked list is empty
        self.__existance_of_elements()

        ## Evaluate the insertion
        if self.head.value == existing_value:
            self.push(value)
        else:
            probe: TwoWayNode = self.head

            while probe != self.tail and probe.value != existing_value:
                probe = probe.next

            if probe.value == existing_value:
                probe.prev.next = TwoWayNode(value, probe, probe.prev)
                probe.prev = probe.prev.next
                self.__count += 1


    def after_value(self, existing_value: int, value: int):
        """Inserts a new node after the first appearance of given value."""
        ## Evaluate the existance
        self.__existance_of_elements()

        if self.count == 1 and self.head.value == existing_value:
            self.append(value)
        else:
            probe: TwoWayNode = self.head

            while probe != self.tail and probe.value != existing_value:
                probe = probe.next

            if probe == self.tail and probe.value == existing_value:
                self.append(value)
            elif probe.value == existing_value:
                probe.next = TwoWayNode(value, probe.next, probe)
                probe.next.next.prev = probe.next
                self.__count += 1


    def index(self, value: int) -> list[int]:
        """Search for the index of the given value and return the list of indixes"""
        return [
            i for i, node in enumerate(self.__iter__()) if node.value == value
        ]


    def clear(self):
        """Deletes the single linked list."""
        self.__head = None
        self.__tail = None
        self.__count = 0


    def pop(self):
        """Deletes the last node of the circular doubly linked list."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        ## If there is only one element
        if self.count == 1:
            self.clear()
        else:
            ## Reassign the tail
            self.__tail = self.tail.prev
            self.__tail.next = self.__head
            self.__count -= 1


    def shift(self):
        """Deletes the first node of the circular doubly linked list."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        ## If there is only one element
        if self.__count == 1:
            self.clear()
        else:
            self.__head = self.__head.next
            self.__head.prev = self.tail
            self.__tail.next = self.head
            self.__count -= 1


    def delete(self, index: int):
        """Deletes and element by an index."""
        ## Verify the index
        self.__verify_index(index)
        
        ## If there is only one element in the list
        if self.count == 1:
            self.clear()
            return

        if index == 0:
            self.shift()
        elif index in [-1, self.count-1]:
            self.pop()
        else:
            probe: TwoWayNode = self.head
            
            for _ in range(index):
                probe = probe.next

            probe.prev.next = probe.next
            probe.next.prev = probe.prev
            self.__count -= 1


    def remove(self, value: int):
        """Deletes the first appearance of the given value."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        if self.head.value == value:
            self.shift()
            return

        probe: TwoWayNode = self.head

        while probe != self.tail and probe.value != value:
            probe = probe.next

        if probe == self.tail and probe.value == value:
            self.pop()
        elif probe != self.tail and probe.value == value:
            probe.prev.next = probe.next
            probe.next.prev = probe.prev
            self.__count -= 1


    def remove_duplicates_sorted(self):
        """Remove the duplicates values in the list."""
        self.__existance_of_elements()

        probe: TwoWayNode = self.head.next
        temp: TwoWayNode = None

        while probe != self.tail:

            if probe.prev.value != probe.value:
                probe = probe.next
            else:
                temp = probe.next
                probe.prev.next = temp
                temp.prev = probe.prev
                self.__count -= 1
                probe = temp

        if probe == self.tail and probe.prev.value == probe.value:
            self.pop()


    def check_sort(self) -> bool:
        """Check if the list is sorted."""
        self.__existance_of_elements()

        x: int -maxsize - 1

        for node in self.__iter__():
            if node.value < x:
                return False

            x = node.value

        return True


    def replace(self, old_value: int,value: int, entire: bool = False):
        """Replaces the first occurrence of the given value with new one or all apperances. By default replace the first appearance."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()
        
        ## Iterate through all the elements in case the all entire parameter is true
        for node in self.__iter__():
            if node.value == old_value:
                node.value = value
                if not entire:
                    break


    def update(self, index: int, value: int):
        """Updates a value in the list by a given index."""
        self.__existance_of_elements()
        ## Evaluates the correct index
        self.__verify_index(index)

        ## Iterate until the given index is reached
        probe: TwoWayNode = self.head
        for _ in range(index):
            probe = probe.next
        
        ## Updates the value
        probe.value = value


    def reverse(self):
        """Reverse the circular doubly linked list backwards."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        probe: TwoWayNode = self.head
        prev_node: TwoWayNode = None
        next_node: TwoWayNode = None

        self.__tail = probe

        for _ in range(self.count):
            next_node = probe.next
            probe.next = prev_node
            probe.prev = next_node
            prev_node = probe
            probe = next_node

        self.__head = prev_node
        self.__head.prev = self.tail
        self.__tail.next = self.head